src/main/scala/org/apache/magellan/GeoJSONRelation.scala

```diff
@@ -0,0 +1,114 @@
+ */
+case class GeoJSONRelation(path: String)
+                          (@transient val sqlContext: SQLContext)
+  extends BaseRelation with TableScan {
+
+  @transient val sc = sqlContext.sparkContext
+
+  override val schema = {
+    StructType(List(StructField("point", new PointUDT(), true),
+      StructField("polyline", new PolyLineUDT(), true),
+      StructField("polygon", new PolygonUDT(), true),
+      StructField("metadata", MapType(StringType, StringType, true), true),
+      StructField("valid", BooleanType, true)
+    ))
+  }
+
+  override def buildScan(): RDD[Row] = {
+    val numFields = schema.fields.length
+    sc.newAPIHadoopFile(
+      path,
+      classOf[WholeFileInputFormat],
+      classOf[NullWritable],
+      classOf[Text]
+    ).flatMap { case(k, v) =>
+      val line = v.toString()
+      parseShapeWithMeta(line)
+    }.mapPartitions { iter =>
+      val row = new GenericMutableRow(numFields)
+      iter.flatMap { case (shape: Shape, meta: Option[Map[String, String]]) =>
+        (0 until numFields).foreach(i => row.setNullAt(i))
+        row(3) = meta.fold(Map[String, String]())(identity)
+        row(4) = if (shape == NullShape) false else shape.isValid()
+        shape match {
+          case NullShape => None
+          case _: Point =>
+            row(0) = shape
+            Some(row)
+          case _: PolyLine =>
+            row(1) = shape
+            Some(row)
+          case _: Polygon =>
+            row(2) = shape
+            Some(row)
+          case _ => ???
+        }
+      }
+    }
+  }
+
+  private def parseShapeWithMeta(line: String) = {
+    val tree = parse(line)
+    implicit val formats = org.json4s.DefaultFormats
+    val result = tree.extract[GeoJSON]
+    result.features.map(f => (f.geometry.shape, f.properties))
+  }
+}
+
+case class Geometry(`type`: String, coordinates: JValue) {
+  def extractPoints(p: List[JValue]) = {
+    p.map { case (JArray(List(JDouble(x), JDouble(y)))) => new Point(x, y)}
+  }
+  val shape = {
+    `type` match {
+      case "Point" => {
+        val JArray(List(JDouble(x), JDouble(y))) = coordinates
+        new Point(x, y)
+      }
+      case "LineString" => {
+        val JArray(p) = coordinates.asInstanceOf[JArray]
+        val points = extractPoints(p)
+        new PolyLine(Array.fill(1)(0), points.toIndexedSeq)
+      }
+      case "Polygon" => {
+        val JArray(p) = coordinates.asInstanceOf[JArray]
+        val rings = p.map { case JArray(q) => extractPoints(q)}
+        val indices = rings.scanLeft(0)((running, current) => running + current.size).dropRight(1)
+        new Polygon(indices.toIndexedSeq, rings.flatten.toIndexedSeq)
+      }
+    }
+  }
+
+}
+case class Feature(`type`: String, properties: Option[Map[String, String]], geometry: Geometry)
+case class CRS(`type`: String, properties: Option[Map[String, String]])
+case class GeoJSON(`type`: String, crs: Option[CRS], features: List[Feature])
```

Comment:
What is CRS? Would you mind adding some documentation on what these various case classes represent? Also, could we rename them to something more specific and/or make them private, since these classes are only used in this file?